src/System.IO.FileSystem/src/System/IO/Win32FileSystemEnumerable.cs

```diff
@@ -513,171 +517,224 @@ private static String GetFullSearchString(String fullPath, String searchPattern)
         {
-            return Win32FileSystemEnumerableHelpers.IsFile(result.FindData);
+            get
+            {
+                if (s_fileInfo == null)
+                {
+                    s_fileInfo = new FileInfoResultHandler();
+                }
+                return s_fileInfo;
+            }
         }
 
-        [System.Security.SecurityCritical]
-        internal override FileInfo CreateObject(SearchResult result)
+        internal static SearchResultHandler<DirectoryInfo> DirectoryInfo
         {
-            String name = result.FullPath;
-            IFileSystemObject fileSystemObject = new Win32FileSystem.Win32FileSystemObject(name, result.FindData, asDirectory: false);
-            FileInfo fi = new FileInfo(name, fileSystemObject);
-            return fi;
+            get
+            {
+                if (s_directoryInfo == null)
+                {
+                    s_directoryInfo = new DirectoryInfoResultHandler();
+                }
+                return s_directoryInfo;
+            }
+        }
+
+        internal static SearchResultHandler<FileSystemInfo> FileSystemInfo
+        {
+            get
+            {
+                if (s_fileSystemInfo == null)
+                {
+                    s_fileSystemInfo = new FileSystemInfoResultHandler();
+                }
+                return s_fileSystemInfo;
+            }
         }
     }
 
-    internal class DirectoryInfoResultHandler : SearchResultHandler<DirectoryInfo>
+    internal class StringResultHandler : SearchResultHandler<string>
     {
-        [System.Security.SecurityCritical]
-        internal override bool IsResultIncluded(SearchResult result)
+        private readonly bool _includeFiles;
+        private readonly bool _includeDirs;
+
+        internal StringResultHandler(bool includeFiles, bool includeDirs)
         {
-            return Win32FileSystemEnumerableHelpers.IsDir(result.FindData);
+            _includeFiles = includeFiles;
+            _includeDirs = includeDirs;
         }
 
         [System.Security.SecurityCritical]
-        internal override DirectoryInfo CreateObject(SearchResult result)
+        internal override bool IsResultIncluded(string fullPath, string userPath, ref Interop.mincore.WIN32_FIND_DATA findData, out string result)
         {
-            String name = result.FullPath;
-            IFileSystemObject fileSystemObject = new Win32FileSystem.Win32FileSystemObject(name, result.FindData, asDirectory: true);
-            DirectoryInfo di = new DirectoryInfo(name, fileSystemObject);
-            return di;
+            bool includeFile = _includeFiles && Win32FileSystemEnumerableHelpers.IsFile(ref findData);
+            bool includeDir = _includeDirs && Win32FileSystemEnumerableHelpers.IsDir(ref findData);
+            Debug.Assert(!(includeFile && includeDir), findData.cFileName + ": current item can't be both file and dir!");
+            if (includeFile || includeDir)
```

Comment:
It's minor, so don't feel obliged to change it, but we're computing includeDir even if includeFile is true.  It could be avoided with either:
```C#
if ((_includeFiles && Win32FileSystemEnumerableHelpers.IsFile(ref findData)) ||
    (_includeDirs && Win32FileSystemEnumerableHelpers.IsDir(ref findData)))
{
    ...
}
```
or something like that.